src/Compilers/Core/Portable/Collections/TopologicalSort.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// A helper class that contains a topological sort algorithm.
+    /// </summary>
+    internal static class TopologicalSort
+    {
+        /// <summary>
+        /// Produce a topological sort of a given directed acyclic graph, given a set of nodes which include all nodes
+        /// that have no predecessors. Any nodes not in the given set, but reachable through successors, will be added
+        /// to the result. This is an iterative rather than recursive implementation, so it is unlikely to cause a stack
+        /// overflow.
+        /// </summary>
+        /// <typeparam name="TNode">The type of the node</typeparam>
+        /// <param name="nodes">Any subset of the nodes that includes all nodes with no predecessors</param>
+        /// <param name="successors">A function mapping a node to its set of successors</param>
+        /// <returns>A list of all reachable nodes, in which each node always precedes its successors</returns>
+        public static ImmutableArray<TNode> IterativeSort<TNode>(IEnumerable<TNode> nodes, Func<TNode, IEnumerable<TNode>> successors)
+        {
+            // First, count the predecessors of each node
+            PooledDictionary<TNode, int> predecessorCounts = PredecessorCounts(nodes, successors);
+
+            // Initialize the ready set with those nodes that have no predecessors
+            var ready = ArrayBuilder<TNode>.GetInstance();
+            foreach ((TNode node, int count) in predecessorCounts)
+            {
+                if (count == 0)
+                {
+                    ready.Push(node);
+                }
+            }
+
+            // Process the ready set. Output a node, and decrement the predecessor count of its successors.
+            var resultBuilder = ImmutableArray.CreateBuilder<TNode>();
+            while (ready.Count != 0)
+            {
+                var node = ready.Pop();
+                resultBuilder.Add(node);
+                foreach (var succ in successors(node))
```

Comment:
💭  I worry a bit about the number of `IEnumerator<T>` allocations required to process the list (potentially 2\**n* for a graph with *n* nodes)

Reply:
We visit each node and edge a constant number of times. We call successors(node) once to count predecessors, and then we call it once on each node to process the outgoing edges. Presuming each call allocates a new enumerator, that is *O(2n)* enumerators.

---
In reply to: [171726056](https://github.com/dotnet/roslyn/pull/25169#discussion_r171726056) [](ancestors = 171726056)